ome/omero-web
# imgData JSON with no stateful services

It has been noted in some cases that the Preview panel and the right panel of iviewer can take a long time to load, since we currently initialise various stateful services (which requires initialising them with the original image files). This means that the viewers are slower to load than if we just get the data from the DB, particularly if access to original data is slow.

Stateful services usage is replaced as follows:

 - Instead of using rendering engine for settings, we load rendering def
 - Instead of the rawPixelsStore for `image.getPixelRange()`, we lookup values based on pixel type
 - Instead of using rendering engine for `requiresPixelsPyramid()` we use `image.requiresPixelsPyramid()` which simply uses the size cut-off.
 - IF the image is BIG, *then* we load the rendering engine to get the actual pyramid resolutions

To test:
 - Image viewer should work as before but load faster
 - The only other change in behaviour is for small images that have a pyramid, such as small NGFF images. These will appear as small images in the Preview panel and will load as whole planes instead of tiles in iviewer.
 - Instead of using `re.lookupRenderingDef(pid, ctx)` for the logic of "use my rdef or use the owners", this is now encoded in `webgateway.util.get_rendering_def()`
 - If no rendering def exists, we use rendering engine to create one as before.

Also see performance testing with NGFF data on idr-testing at https://github.com/ome/omero-web/pull/537#issuecomment-2004001921
Summarised as follows:
 - Don't expect this PR to noticeably improve performance for *most* data in OMERO
 - For data where we do notice a significant improvement (e.g. idr0090 NGFF data) even if we always use RenderingEngine to test/load the pyramid levels (ie. NO api change), this PR still has a performance improvement.
